src/span.rs

```diff
@@ -0,0 +1,24 @@
+use std::ops::Range;
+
+// A source code span, used for error reporting.
+#[derive(Clone)]
+pub enum Span {
```

Comment:
do we need an entirely new module for this? maybe it can live in `lib.rs`

Reply:
Originally I planned to put span derivation impls in here, but so far they've been unnecessary. I propose we merge this into lib.rs if by the end of prototype this turns out to be sufficiently smol.

Reply:
Looks like we're at end. IMO a separate module is fine, since Span tries to be a semi-opaque type (even in frontend, at a certain level you just get, clone and merge spans). Wyt?